pkg/kubelet/stats/cri_stats_provider.go

```diff
@@ -57,10 +65,203 @@ func newCRIStatsProvider(
+	if err != nil {
+		return nil, fmt.Errorf("failed to list all pod sandboxes: %v", err)
+	}
+	for _, s := range podSandboxes {
+		podSandboxMap[s.Id] = s
+	}
+
+	// uuidToFsInfo is a map from filesystem UUID to its stats. This will be
+	// used as a cache to avoid querying cAdvisor for the filesystem stats with
+	// the same UUID many times.
+	uuidToFsInfo := make(map[runtimeapi.StorageIdentifier]*cadvisorapiv2.FsInfo)
+
+	// sandboxIDToPodStats is a temporary map from sandbox ID to its pod stats.
+	sandboxIDToPodStats := make(map[string]*statsapi.PodStats)
+
+	resp, err := p.runtimeService.ListContainerStats(&runtimeapi.ListContainerStatsRequest{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to list all container stats: %v", err)
+	}
+	for _, stats := range resp.Stats {
+		containerID := stats.Attributes.Id
+		container, found := containerMap[containerID]
+		if !found {
+			glog.Errorf("Unknown id %q in container map: %v", containerID)
+			continue
+		}
+
+		podSandboxID := container.PodSandboxId
+		podSandbox, found := podSandboxMap[podSandboxID]
+		if !found {
+			glog.Errorf("Unknown id %q in pod sandbox map: %v", podSandboxID)
+			continue
+		}
+
+		// Creates the stats of the pod (if not created yet) which the
+		// container belongs to.
+		ps, found := sandboxIDToPodStats[podSandboxID]
+		if !found {
+			ps = p.makePodStats(podSandbox)
+			sandboxIDToPodStats[podSandboxID] = ps
+		}
+		ps.Containers = append(ps.Containers, *p.makeContainerStats(stats, container, &rootFsInfo, uuidToFsInfo))
+	}
+
+	result := make([]statsapi.PodStats, 0, len(sandboxIDToPodStats))
+	for _, s := range sandboxIDToPodStats {
+		result = append(result, *s)
+	}
+	return result, nil
 }
 
+// ImageFsStats returns the stats of the image filesystem.
 func (p *criStatsProvider) ImageFsStats() (*statsapi.FsStats, error) {
-	return nil, fmt.Errorf("not implemented")
+	resp, err := p.imageService.ImageFsInfo(&runtimeapi.ImageFsInfoRequest{})
+	if err != nil {
+		return nil, err
+	}
+
+	for _, fs := range resp.ImageFilesystems {
+		s := &statsapi.FsStats{
+			Time:       metav1.NewTime(time.Unix(0, fs.Timestamp)),
+			UsedBytes:  &fs.UsedBytes.Value,
+			InodesUsed: &fs.InodesUsed.Value,
+		}
+		imageFsInfo := p.getFsInfo(fs.StorageId)
+		if imageFsInfo != nil {
+			s.AvailableBytes = &imageFsInfo.Available
+			s.CapacityBytes = &imageFsInfo.Capacity
+			s.InodesFree = imageFsInfo.InodesFree
+			s.Inodes = imageFsInfo.Inodes
+		}
+		return s, nil
+	}
+
+	return nil, fmt.Errorf("imageFs information is unavailable")
+}
+
+// getFsInfo returns the information of the filesystem with the specified
+// storageID. If any error occurs, this function logs the error and returns
+// nil.
+func (p *criStatsProvider) getFsInfo(storageID *runtimeapi.StorageIdentifier) *cadvisorapiv2.FsInfo {
+	if storageID == nil {
+		glog.V(2).Infof("Failed to get filesystem info: storageID is nil")
+		return nil
+	}
+	fsInfo, err := p.cadvisor.GetFsInfoByFsUUID(storageID.Uuid)
```

Comment:
Is omitting the filesystem info on cadvisor error the current behavior or a new one introduced in this PR (due to the storage id mapping)?

Reply:
It's only here in this PR.

Reply:
When will `GetFsInfoByFsUUID` return an error (assuming the uuid exists)?

Reply:
`GetFsInfoByFsUUID` and `GetDirFsInfo` (used for retrieving the rootFs information from cAdvisor) convert the UUID and the directory name to the device name and call `getFsInfoByDeviceName` to get the filesystem stats. So `GetFsInfoByFsUUID` can fail with the same reasons as `GetDirFsInfo` could fail.

Do you suggest we return the error instead of partial summary when the UUID does exist but something wrong when fetching the stats?

Reply:
Add a comment to explain we return partial results. Also check (or file an issue against) heapster to make sure 1) it is capable of processing partial results, and 2) the monitoring pipeline prefers partial results.